https://thefiddler.substack.com/p/can-you-see-between-the-trees

---

Um, I can't think of any clever way to solve these puzzles, so I am just going to write some code.

---

In [71]:
import math as m

# V is the visibility (distance that you can see)
def find_sorted_gaps(V, full_quadrant=True):
    pts = []
    V_sq = V*V
    for x in range(V+1):
        x_sq = x*x
        for y in range((V if full_quadrant else x)+1):
            if (x != 0 or y != 0):
                y_sq = y*y
                if (x_sq + y_sq <= V_sq):
                    if (m.gcd(x,y) == 1):
                        # Found a valid visible point.
                        a = m.degrees(m.atan2(y,x))
                        pts.append((a,x,y))
    
    pts.sort()
    l_pts = len(pts)

    gaps = []
    for i in range(l_pts-1):
        a1,x1,y1 = pts[i]
        a2,x2,y2 = pts[i+1]
        g = a2 - a1
        gaps.append((g,a1,a2,x1,y1,x2,y2))

    gaps.sort(reverse=True)

    return gaps

find_sorted_gaps(4, full_quadrant=True)

[(18.43494882292201, 71.56505117707799, 90.0, 1, 3, 0, 1),
 (18.43494882292201, 0.0, 18.43494882292201, 1, 0, 3, 1),
 (11.309932474020215, 45.0, 56.309932474020215, 1, 1, 2, 3),
 (11.309932474020215, 33.690067525979785, 45.0, 3, 2, 1, 1),
 (8.13010235415598, 63.43494882292201, 71.56505117707799, 1, 2, 1, 3),
 (8.13010235415598, 18.43494882292201, 26.56505117707799, 3, 1, 2, 1),
 (7.125016348901795, 56.309932474020215, 63.43494882292201, 2, 3, 1, 2),
 (7.125016348901795, 26.56505117707799, 33.690067525979785, 2, 1, 3, 2)]

Looks good so far. Matches up with the example provided.

In [72]:
def top_gap_locs(V, idx_lo=0, idx_hi=20, full_quadrant=False):
    gaps = find_sorted_gaps(V, full_quadrant=full_quadrant)
    locs = []
    for i in range(idx_lo, idx_hi + 1):
        if (i < len(gaps)):
            g = gaps[i]
            sz, a1, a2 = g[0], g[1], g[2]
            #a_avg = (a1+a2)/2
            locs.append(f"{a1=}, {a2=}, {sz=}")

    locs.sort()
    return locs

----

Let's use that code to find the solution to the Fiddler, which wants the largest pair of gap in the range 0 to 45 after the ones next to 0 and 45.

In [73]:
for V in range(20,200,10):
    print(top_gap_locs(V,2,3))

['a1=25.20112364547507, a2=26.56505117707799, sz=1.3639275316029185', 'a1=26.56505117707799, a2=27.89727103094763, sz=1.3322198538696384']
['a1=25.709953780811265, a2=26.56505117707799, sz=0.855097396266725', 'a1=26.56505117707799, a2=27.474431626277134, sz=0.9093804491991442']
['a1=25.906507999514385, a2=26.56505117707799, sz=0.6585431775636046', 'a1=26.56505117707799, a2=27.216111557307478, sz=0.6510603802294881']
['a1=26.02959219151346, a2=26.56505117707799, sz=0.5354589855645315', 'a1=26.56505117707799, a2=27.0955524937518, sz=0.53050131667381']
['a1=26.13100054493858, a2=26.56505117707799, sz=0.4340506321394102', 'a1=26.56505117707799, a2=26.99583839408662, sz=0.43078721700862843']
['a1=26.188110697480855, a2=26.56505117707799, sz=0.3769404795971347', 'a1=26.56505117707799, a2=26.939528063800815, sz=0.37447688672282453']
['a1=26.241349652157766, a2=26.56505117707799, sz=0.3237015249202244', 'a1=26.56505117707799, a2=26.886934192476716, sz=0.32188301539872555']
['a1=26.274212154727

From the output above (and below), we can see that the desired pair of gaps is around the angle 26.565 degrees.

**And the answer to the fiddler is 26.56505117707799 degrees (corresponding to the tree at point x=2,y=1)**

---

The extra credit: "The fifth largest pair of adjacent gaps in this range are on either side of what angle up from the x-axis? "
So, let's first study the output to make sure that things are paired as we expect.

In [74]:
gg = find_sorted_gaps(100, full_quadrant=False)
gg[0:20]

[(0.5787255656077621, 0.0, 0.5787255656077621, 1, 0, 99, 1),
 (0.4063462333090371, 44.59365376669096, 45.0, 71, 70, 1, 1),
 (0.25808735133618654, 26.306963825741803, 26.56505117707799, 89, 44, 2, 1),
 (0.2569300245035109, 26.56505117707799, 26.8219812015815, 2, 1, 89, 45),
 (0.18663054356556685, 18.43494882292201, 18.621579366487577, 3, 1, 92, 31),
 (0.1830529858434673, 18.251895837078543, 18.43494882292201, 94, 31, 3, 1),
 (0.1609427775784127, 33.690067525979785, 33.8510103035582, 3, 2, 82, 55),
 (0.15959785884668776, 33.5304696671331, 33.690067525979785, 83, 55, 3, 2),
 (0.1418209467804239, 14.036243467926479, 14.178064414706903, 4, 1, 95, 24),
 (0.13906715291111915, 13.89717631501536, 14.036243467926479, 97, 24, 4, 1),
 (0.11887074129995767, 36.86989764584402, 36.98876838714398, 4, 3, 77, 58),
 (0.11716911992492385, 11.309932474020213, 11.427101593945137, 5, 1, 94, 19),
 (0.11621846031255245, 36.75367918553147, 36.86989764584402, 79, 59, 4, 3),
 (0.11482104772023582, 11.195111426299

Things do seem to pair up nicely.

But the pair around 36.87 and 11.31 seem to be very close in size, and it is not clear which is larger. Let's see if things are clearer with larger V.

In [75]:
gg2 = find_sorted_gaps(1000, full_quadrant=False)
gg2[10:14]

[(0.011500557756647822, 36.86989764584402, 36.88139820360067, 4, 3, 797, 598),
 (0.011475221058880436, 36.85842242478514, 36.86989764584402, 799, 599, 4, 3),
 (0.01128980862493556, 11.298642665395278, 11.309932474020213, 976, 195, 5, 1),
 (0.011254327003772602,
  11.309932474020213,
  11.321186801023986,
  5,
  1,
  979,
  196)]

Trying out various larger values of V (200, 500, 1000, 10000), it is clear that the gap around 36.8698 is larger than the gap around 11.3099. 

**And so, the answer to the extra credit is 36.86989764584402 degrees (corresponding to the tree at point x=4,y=3).**

----